Post resourses

td-extreme opened the pull request:
Created `HanlderPostFileContents` class.

Added that handler to the `HanlderRouter`.

Updated `IMessageIO` interface and `SocketIO` class with method to get a
byte[] stream

Modified `HttpRequestBuilder` so that instead of being supplied a String
containing a rawRequest, it gets passed a reference to the `SocketIO`
class. This way after parsing the request line and headers it can check
to see if it needs to get a body.

`HttpReaderWriter` updated to work with the changes made to
`HttpRequestBuilder`.

`HttpRequest` now stores a byte[] body, HashMap of headers, and has a
method to get the ContentLength.

Updated `MockFileIO` for use in tests.

td-extreme committed:
Ability to write to files

td-extreme committed:
Feature POST resources

Created HanlderPostFileContents class.

Added that handler to the HanlderRouter.

Updated IMessageIO interface and SocketIO class with method to get a
byte[] stream

Modified HttpRequestBuilder so that instead of being supplied a String
containing a rawRequest, it gets passed a reference to the SocketIO
class. This way after parsing the request line and headers it can check
to see if it needs to get a body.

HttpReaderWriter updated to work with the changes made to
HttpRequestBuilder.

HttpRequest now stores a byte[] body, HashMap of headers, and has a
method to get the ContentLength.

Updated MockFileIO for use in tests.

td-extreme committed:
Return code 422 for POST with no body.

td-extreme committed:
removed debug code

td-extreme committed:
tests for HandlerRouter

mikeknep commented on src/main/java/handlers/HandlerPostFileContents.java:
```diff
@@ -0,0 +1,57 @@
+  private String createDuplicateFile(String path, int duplicateFileIndex) {
+    int extensionPosition = path.lastIndexOf('.');
+    String fileName;
+    String fileExtention;
```
nit, typo, `Extension`

mikeknep commented on src/main/java/io/HttpReaderWriter.java:
```diff
@@ -15,8 +15,7 @@ public HttpReaderWriter(IMessageIO messageIO, IRequestBuilder builder, HttpRespo
     socketIO.openClientConnection();
-    String message = getMessage();
-    return builder.createRequest(message);
+    return builder.createRequest(socketIO);
```
I find it a little strange that the `HttpReaderWriter` no longer knows how to read the HTTP request and instead is delegating that responsibility to the `RequestBuilder`. Working with sockets seems like a pretty significant enough responsibility in its own right; adding it to "parsing the message from that socket" might be an SRP violation.

td-extreme replied:
Yeah, I wasn't totally sure how to go about this. My other option was to keep it the way it was and then have the HttpReaderWriter check to see if the `Request` returned by the `HttpRequestBuilder` had a content length and then get the body and add it to the `Request` But then it seemed liked the `HttpReaderWriter` would be doing more than it should.

mikeknep commented on src/test/java/HandlerRouterTest.java:
```diff
@@ -0,0 +1,60 @@
+import com.td.Mocks.*;
+import java.util.HashMap;
+
+public class HandlerRouterTest extends junit.framework.TestCase {
```
👍 👍

td-extreme committed:
Fixed typo

td-extreme committed:
move SocketIO use out of HttpRequestBuilder

dadarek commented on src/main/java/interfaces/IRequestBuilder.java:
```diff
@@ -1,5 +1,6 @@
 
 public interface IRequestBuilder {
-  public HttpRequest createRequest(String rawRequest) throws InvalidHttpRequestException;
+  public HttpRequest createRequest(String rawRequest) throws IOException, InvalidHttpRequestException;
```
Why are we all of a sudden exposing the low level `IOException`?

Two things about this:
1. Ideally, anything using this interface only has to worry about exceptions that are at the same level as this interface. `IOException` is certainly below that.
2. It's very suspicious that an _interface_ has to change to all of a sudden declare that it may throw this exception. Seems like the interface is getting coupled to the thing that implements this interface, which should _never_ be the case. If an interface changes, it's because the client of the interface requires it to change.
3. Fill in the blank: "Point # 2 above is the _ in SOLID"

td-extreme replied:
I want to talk about this during our standup tomorrow afternoon.

dadarek commented on src/main/java/interfaces/IFileIO.java:
```diff
@@ -5,6 +5,8 @@
   public String workingDirectory();
   public byte[] getContent(String fileName) throws IOException;
   public String[] getFiles(String directory) throws IOException;
+  public void writeContent(String fileName, byte[] body) throws IOException;
```
Ever heard of the I in SOLID?

This interface reminds of that :) Although these are exclusively file/directory operators.

Can you answer the question "who owns this interface"?

dadarek commented on src/main/java/handlers/HandlerPostFileContents.java:
```diff
@@ -0,0 +1,57 @@
+      fileName = path.substring(0, extentionPosition);
+      fileExtention = path.substring(extentionPosition);
+    }
+    return fileName + "(" + duplicateFileIndex + ")" + fileExtention;
```
Also, minor thing, but I'd avoid parens in filenames. It just gets hairy. Stick with periods or dashes or underscores.

i.e. myfile.1.txt

td-extreme replied:
Okay. Changing this to use a dash instead

dadarek commented on src/main/java/handlers/HandlerPostFileContents.java:
```diff
@@ -0,0 +1,57 @@
+    return rtnFile;
+  }
+
+  private String createDuplicateFile(String path, int duplicateFileIndex) {
```
This doesn't actually "create" a file. The name is misleading.

td-extreme replied:
Okay, this should probably be named.  `createDuplicateFileName` or actually `generateDuplicateFileName`

dadarek commented on src/main/java/handlers/HandlerPostFileContents.java:
```diff
@@ -0,0 +1,57 @@
+
+  private String getFileToWrite(String path) throws IOException {
+    if (!fileIO.exists(path)) {
+      return path;
```
Not a big deal - but in general it's good practice to have just one return statement. Multiple points of exit get confusing.

_Not_ a big problem for small methods, but ...

td-extreme replied:
Okay. I do see a way to refactor this so that will remove the if check and have only one exit point.

dadarek commented on src/main/java/request/HttpRequest.java:
```diff
@@ -1,13 +1,22 @@
+  public HttpRequest(String requestLine, HashMap<String, String> headers) {
+    this.requestLine = requestLine;
+    this.headers = headers;
+    this.body = "".getBytes();
```
Can't this just be `new byte[0]` ?

dadarek commented on src/main/java/request/HttpRequest.java:
```diff
@@ -1,13 +1,22 @@
     this.requestLine = requestLine;
   }
 
+  public HttpRequest(String requestLine, HashMap<String, String> headers) {
```
Does this new constructor eliminate the need for the existing 1-argument ctor?

td-extreme replied:
Yes, that constructor is now useless.

dadarek commented on src/main/java/request/HttpRequest.java:
```diff
@@ -19,7 +28,29 @@ public String method() {
     return requestLine().split(" ")[2];
   }
+
+  public String getValueForHeader(String key) {
```
Why is this public?

dadarek commented on src/main/java/request/HttpRequest.java:
```diff
@@ -19,7 +28,29 @@ public String method() {
+    String length = getValueForHeader("Content-Length");
+    int rtnLength = 0;
+    if (length != null) {
+      rtnLength = Integer.parseInt(length);
```
Although we're not going to be running business-critical things on this server, it _might_ be good to start thinking about "how will an external system break my code".

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -9,12 +10,33 @@ public HttpRequestBuilder(IValidator requestVerifier) {
-  public HttpRequest createRequest(String rawRequest) throws InvalidHttpRequestException {
-    String rawRequestArray[] = rawRequest.split("\n", 2);
+  public HttpRequest createRequest(String rawRequest) throws IOException, InvalidHttpRequestException {
+    HttpRequest rtnRequest;
```
How about just `result` or `request` so we avoid the awkward abbreviation.

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -9,12 +10,33 @@ public HttpRequestBuilder(IValidator requestVerifier) {
+    return rtnRequest;
+  }
+
+  public String parseRequestLine(String rawRequest) throws InvalidHttpRequestException {
```
Let's be a little more diligent with the choice of visibility modifiers.

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -9,12 +10,33 @@ public HttpRequestBuilder(IValidator requestVerifier) {
+    return requestLine;
+  }
+
+  public HashMap<String, String> parseHeaders(String rawRequest) throws InvalidHttpRequestException  {
```
visibility modifier

dadarek commented on src/main/java/request/HttpRequestBuilder.java:
```diff
@@ -9,12 +10,33 @@ public HttpRequestBuilder(IValidator requestVerifier) {
+    String headerLines[] = rawHeaders.split("\n");
+    for (String line : headerLines ) {
+      if (line.contains(":") && line.contains(" ")) {
+        String key = line.split(":", 2)[0];
```
Can we do a `split(": ")` in java (with the whitespace already included in the split-by-string? Would that avoid this weird double-split?

Also, could probably do the split right away and then have a `if (splitThing.length == 2)` instead of the weird `contains` logic.

(I'm not actually sure that you can split by a multi-character string.)

dadarek commented on src/test/java/HandlerRouterTest.java:
```diff
@@ -0,0 +1,60 @@
+  protected void setUp() {
+    mockFileIO = new MockFileIO();
+    handlerRouter = new HandlerRouter(mockFileIO);
+    request = new HttpRequest("GET /file.txt HTTP/1.1", new HashMap<String, String>());
```
It's a little weird that the setup creates a `GET` request that's only used by half of the tests.

dadarek commented on src/test/java/HandlerRouterTest.java:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  public void testRouterForPostWithOutBody() {
+    request = new HttpRequest("POST /file.txt HTTP/1.1", new HashMap<String, String>());
```
All this duplication!

td-extreme committed:
Duplicate file name uses - instead of ()

Also changed name of method from createDuplicateFile to
generateDuplicateFileName.

td-extreme committed:
changed .getBytes() to new byte[0]

td-extreme committed:
Remove dead constructor

td-extreme committed:
Changed getValueForHeader to private, removed test for it

td-extreme committed:
Refactored method for getting content length.

Removed the getValueForHeader method.

Removed the check for null and instead used the getOrDefault method
to return 0 if there is not a Content-Length header.

td-extreme committed:
Fix visiblity modifiers, remove IOException, abbr.

Changed methods that should be private form public to private.

Removed the throws IOExcpetion form IRequestBuilder interface and
from HttpRequestBuilder.

Removed rtn abbreviation from retrun variables.

td-extreme committed:
Refactored method to make it easier to read.

td-extreme committed:
Removed duplication.

td-extreme committed:
Cleaned up parsing of headers
